src/services/goals.js

```diff
@@ -2571,3 +2572,140 @@ export async function createMultiRecipientGoalsFromAdmin(data) {
     grantsForWhichGoalWillBeCreated: [],
   };
 }
+
+const fieldMappingForDeduplication = {
+  name: 'name',
+  source: 'source',
+  status: 'status',
+  responsesForComparison: 'responsesForComparison',
+};
+
+/**
+ *
+ * similarity response is an array of objects with the following structure:
+ * {
+      "id": number,
+      "name": string,
+      "matches": [{
+        "id": number,
+        "name": string,
+        "grantId": number,
+        "similarity": float,
+      }]
+   }
+ *
+ * @param {[]} similarityResponse
+ * @returns {
+ *  goals: [{
+ *    name: string,
+ *    source: string,
+ *    status: string,
+ *    responsesForComparison: string,
+ *    ids: number[],
+ *  }],
+ *  ids: number[]
+ * }[]
+ */
+export async function getGoalIdsBySimilarity(similarityResponse) {
+  // convert the response to a group of IDs
+  const goalIdGroups = similarityResponse.map((matchedGoals) => {
+    const { id, matches } = matchedGoals;
+    return uniq([id, ...matches.map((match) => match.id)]);
+  });
+
+  // convert the ids to a big old database query
+  const goalGroups = await Promise.all(goalIdGroups.map((group) => Goal.findAll({
+    attributes: ['id', 'status', 'name', 'source', 'goalTemplateId', 'grantId'],
+    where: {
+      [Op.or]: [
+        {
+          id: group,
+          '$"goalTemplate"."creationMethod"$': {
+            [Op.is]: null,
+          },
+        },
+        {
+          id: group,
+          '$"goalTemplate"."creationMethod"$': {
+            [Op.eq]: CREATION_METHOD.CURATED,
+          },
+          status: {
+            [Op.not]: GOAL_STATUS.CLOSED,
+          },
+        },
+      ],
+    },
+    include: [
+      {
+        model: Grant,
+        as: 'grant',
+        required: true,
+        attributes: ['id', 'status'],
+      },
+      {
+        model: GoalFieldResponse,
+        as: 'responses',
+        required: false,
+        attributes: ['response', 'goalId'],
+      },
+      {
+        model: GoalTemplate,
+        as: 'goalTemplate',
+        required: false,
+        attributes: ['id', 'creationMethod'],
+      },
+    ],
+  })));
+
+  const uniqueGrantIds = uniq(goalGroups.map((group) => group.map((goal) => goal.grantId)).flat());
+
+  const grants = await Grant.findAll({
+    where: {
+      [Op.or]: [
+        { id: uniqueGrantIds },
+        { oldGrantId: uniqueGrantIds },
+      ],
+      status: 'Active',
+    },
+    attributes: ['id', 'oldGrantId', 'status'],
+  });
+
+  const grantLookup = {};
+  grants.forEach((grant) => {
+    grantLookup[grant.id] = grant.id;
+    grantLookup[grant.oldGrantId] = grant.id;
```

Comment:
this could be a null key, as not all grants have an oldGrantId

Reply:
that's ok, ```grantLookup[current.grantId]``` will be falsy still